### PR TITLE
User-definable custom module support in model YAML

### PR DESCRIPTION
**Pull request by Y-T-G**
Since there are users looking for more customization when it comes to custom modules (https://github.com/ultralytics/ultralytics/pull/19609, https://github.com/ultralytics/ultralytics/pull/18909), this PR provides an alternative that enables a customizable interface for users to define and use custom modules without modifying Ultralytics source code. It takes inspiration from the `download` script feature that Ultralytics utilizes for dataset YAML.

For a user to define a custom module, they need to simply add the definition code and optionally the parser code to the model YAML as string.

In this YAML, I have defined two example modules directly in YAML, `Backbone` and `Head`. They don't exist in Ultralytics. In Ultralytics, we need to typically do two things to add a module. First, define the module and then if needed [specify a custom parser](https://github.com/ultralytics/ultralytics/blob/da98efc61d9e0467315fc86c2297c8d81e656b1a/ultralytics/nn/tasks.py#L1679) in `tasks.py` to parse the arguments from the YAML for that module.

Both of these can be done directly in the YAML with this PR. The `init` section in YAML is where you would define the modules. You can define any number of modules here. For example, I defined two different modules `Backbone` and `Head`. And then in the `parse` section you define the parser (if needed):

```yaml
 nc: 10
 backbone:
   - [-1, 1, Backbone, []]  # uses Backbone Module defined below
 head:
   - [0, 1, Head, [1, nc]]  # uses Head Module defined below
 
 module: 
   init: |
     # Define all custom modules here.
     # Make sure the indentations are correct.
 
     import torch.nn as nn
 
     class Backbone(nn.Module):
         def __init__(self):
             super().__init__()
             self.backbone = nn.Sequential(
                 nn.Conv2d(3, 16, kernel_size=3, stride=1, padding=1),  # (3,640,640) -> (16,640,640)
                 nn.ReLU(),
                 nn.MaxPool2d(2, 2),  # (16,640,640) -> (16,320,320)
                 nn.Conv2d(16, 32, kernel_size=3, stride=1, padding=1),  # (16,320,320) -> (32,320,320)
                 nn.ReLU(),
                 nn.MaxPool2d(2, 2),  # (32,320,320) -> (32,160,160)
                 nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=1),  # (32,160,160) -> (64,160,160)
                 nn.ReLU(),
                 nn.MaxPool2d(2, 2)   # (64,160,160) -> (64,80,80)
             )
 
         def forward(self, x):
             x = self.backbone(x)
             return x
         
     class Head(nn.Module):
         def __init__(self, num_classes=10):
             super().__init__()
             self.head = nn.Sequential(
                 nn.Flatten(),  # Flatten (64,80,80) -> (64*80*80)
                 nn.Linear(64 * 80 * 80, 128),
                 nn.ReLU(),
                 nn.Linear(128, num_classes)
             )
 
         def forward(self, x):
             x = self.head(x)
             return x
         
   parse: |
     # parser for args; to modify input arguments if needed
     if m is Head:
       c2 = args[0]  # first arg in YAML is output channel
       c1 = ch[f]  # automatically get input channel from input layer
       args = [*args[1:]]  # skip first arg that was provided in YAML and pass the rest to __init__()
```

Then we simply load the YAML in Ultralytics:
```python
 In [8]: model = YOLO("custom-classifier.yaml", task="classify")
 custom-classifier summary (fused): 13 layers, 52,453,802 parameters, 52,453,802 gradients, 2.3 GFLOPs
```

And use it as usual:
```python
 In [9]: results = model.train(data="imagenet10", epochs=1, verbose=False, exist_ok=True, imgsz=640)
 train: /datasets/imagenet10/train... found 12 images in 10 classes ✅ 
 val: /datasets/imagenet10/val... found 12 images in 10 classes ✅ 
 test: None...
 
                    from  n    params  module                                       arguments                     
   0                  -1  1     23584  ultralytics.nn.tasks.Backbone                []                            
   1                   0  1  52430218  ultralytics.nn.tasks.Head                    [10]                          
 custom-classifier summary (fused): 13 layers, 52,453,802 parameters, 52,453,802 gradients, 2.3 GFLOPs
 Starting training for 1 epochs...
 
       Epoch    GPU_mem       loss  Instances       Size
         1/1      2.66G      2.297         12        640: 100%|██████████| 1/1 [00:02<00:00,  2.20s/it]
                classes   top1_acc   top5_acc: 100%|██████████| 1/1 [00:00<00:00,  8.69it/s]
                    all     0.0833      0.417
 
 1 epochs completed in 0.002 hours.
 
 Validating /ultralytics/runs/classify/train/weights/best.pt...
 test: None...
                classes   top1_acc   top5_acc: 100%|██████████| 1/1 [00:00<00:00, 40.88it/s]
                    all     0.0833      0.417
 Speed: 0.7ms preprocess, 0.6ms inference, 0.0ms loss, 0.0ms postprocess per image
 Results saved to /ultralytics/runs/classify/train
```

The parser has access to the local and global namespace. Although the local namespace isn't mutable, the global namespace is, which enables `init` code to add custom module into the namespace.

This will hopefully meet the requirements of the users, whilst keeping the Ultralytics codebase free of additional dependencies. It enables flexibility while avoiding bloat.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds support for custom parsing hooks in model configs, enabling advanced users to inject Python code to initialize and modify how models are built at parse time. 🧩⚙️

### 📊 Key Changes
- Introduces two optional hooks in the model dict:
  - `module.init`: executed once before model parsing to run custom initialization.
  - `module.parse`: executed inside the parse loop to dynamically adjust layer construction (e.g., override `args`).
- Implements safe fallbacks: if hooks are absent, nothing changes (backward compatible).
- Scope handling:
  - `module.init` runs with `globals()`.
  - `module.parse` runs with `globals()` and a local `namespace`, and can modify `args` via `namespace["args"]`.

Example usage in a model YAML/dict:
```yaml
module:
  init: |
    from my_blocks import MyBlock  # register/import custom modules

  parse: |
    # Example: dynamically modify args for a specific module
    if m == 'MyBlock':
        # c1, c2, f, n, stride, etc. are available in scope
        args = [c1, c2, 3, 1]  # override args for this layer
```

### 🎯 Purpose & Impact
- Enables powerful extensibility without modifying Ultralytics source code, supporting custom layers, dynamic architectures, and research workflows. 🚀
- Maintains full backward compatibility; existing models behave unchanged. ✅
- Potential risks/considerations:
  - Security: executing arbitrary code in configs—only load trusted models. 🔐
  - Reproducibility: custom code paths may affect determinism; document hooks and seed settings as needed. 📦
- Improves portability of custom models across environments (custom logic embedded in the config). 🌍

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

Your PR introduces an exciting new feature with custom module initialization and parsing! 🎉 To make this enhancement even better:
- 🔍 If applicable, please provide a **minimum reproducible example (MRE)** that demonstrates the use of this new functionality. This will assist us in reviewing and testing the changes effectively.  
- 🧪 Ensure that any new functionality is tested adequately, including edge cases, to maintain reliability.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). This is an automated response, but don’t worry—an Ultralytics engineer will review your PR and provide further assistance soon. Thank you for contributing to Ultralytics! 🚀

**Comment by xceccon** [comment]
Hello YTG

I have a question about this PR.

Can I define more heads ? Can I define let's say 6 heads in yolov11 instead of the default 3 heads ?

I would like to experiment because I think that there is a small probability that having a lot of different specialised heads can be like ensembling full yolo models with a fraction of the cost.

**Comment by Y-T-G** [comment]
You can as long as the output is in similar format to the original and you inherit from the `Detect` class.

File: ultralytics/nn/tasks.py [hunk 1]
```diff
@@ -1631,6 +1631,7 @@ def parse_model(d, ch, verbose=True):
             A2C2f,
         }
     )
+    exec(d.get("module", {}).get("init", ""), globals())  # run custom init script
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py]: ❗ **CRITICAL**: Executing arbitrary strings from the YAML with `exec(...)` turns every model YAML load into a remote code execution vector. Any malicious or tampered YAML (e.g., downloaded from the internet or supplied by a dataset) can run unrestricted Python on the user’s machine long before the model is validated. Please gate this behind an explicit, opt-in safety flag, or otherwise sandbox/validate the code before execution, and apply the same mitigation to the parser `exec` below.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Blocking: Running YAML-defined Python via `exec` introduces an immediate remote code execution vulnerability. Please add an explicit opt-in or harden the execution path before proceeding.

💬 Posted 1 inline comment

File: ultralytics/nn/tasks.py [hunk 2]
```diff
@@ -1631,6 +1631,8 @@ def parse_model(d, ch, verbose=True):
             A2C2f,
         }
     )
+    if SETTINGS.get("yaml_exec"):
+        exec(d.get("module", {}).get("init", ""), globals())  # run custom init script
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py:1607]: ⚠️ **HIGH**: Thanks for adding the opt-in guard, but calling `d.get("module", {}).get(...)` assumes the YAML `module` entry is always a dict. If a user accidentally sets `module` to a string or list (e.g., a script path), this will raise an `AttributeError` before you even check the flag. Could you normalize/validate the value first (and apply the same safeguard in the parser block below) so the code fails gracefully?

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please add type checks around the new YAML `module` handling so we don’t crash when users provide a non-dict value.

💬 Posted 1 inline comment

**Comment by glenn-jocher** [comment]
Thank you @Y-T-G for addressing the real demand for custom model modules and extensibility. We’re closing this design because executing Python strings embedded in model YAML crosses a security boundary and bypasses the module-registry ownership that keeps configurations portable and reviewable. The use case remains valuable, but it should be pursued through an explicit registry/plugin or safe import mechanism rather than `exec`. We appreciate the concise proposal and the discussion it enabled. 🙏
